ls1intum/Artemis
# [Feature] automatic text/feedback consistency ui

### Checklist

### Motivation and Context


It displays feedback conflicts in text assessments to the tutors/instructors. It allows them to see/compare two feedback side by side and update the conflicting one.
### Description


- Added one boolean field to FeedbackConflict entity (db change)
- Added one new REST call for getting submissions with conflicting feedback
- Added one new REST call for manually solving feedback conflict
- Added one new page for displaying multiple submissions 

### Steps for Testing


- You can use the course and text exercise that I created in [test server 3](https://vmbruegge60.in.tum.de/) and skip to the step 5.

- If you want to test this locally, you have to run the Athene as well.

1. Log in to Artemis
2. Create a text exercise with automatic assessments are enabled.
3. Create 5-6 submissions.
4. Wait for the submission due date so clusters can be generated.
5. Start assessing. (better to use different tutors)
6. Look for the feedback with automatic suggestions and assess it completely differently. (Submit not save)
7. If you wait for few seconds and refresh the page you will see the conflict badge. (Screenshot 1)
8. Click on the badge and navigate to conflict page. (GIF 1)
9. You can see the yellow conflicting feedback pairs and switch to other submissions if there are any.
10. You can deal with the conflict in two ways
      a) Change the feedback on the left and override the submission. (GIF 2)
           - With this approach, conflict might be solved by Athene or more conflicts can be generated.
           - Clicking this will take you to the previous page, you have to wait and refresh the page to see if the conflict is resolved. (Athene calculation works asynchronously)
       b) You can select one of the conflicting feedback on the right and mark it as no conflict(button name changed into -> discard). (GIF 3)
            - which means according to the tutor that is not a conflict.
            - You can select only one conflicting feedback on the right. (In theory, there can be more than conflicts)
            - This will also take you to the previous page, you should see number of conflicts are reduced for that feedback.

### Screenshots

<img width="1425" alt="Conflict badge" src="https://user-images.githubusercontent.com/60740930/95020251-59556200-066a-11eb-9fa0-4889fb968d17.png">


![GIF 1](https://user-images.githubusercontent.com/60740930/95021961-7000b680-0674-11eb-85d5-0bd2f08e088d.gif)

![GIF 2](https://user-images.githubusercontent.com/60740930/95021962-7131e380-0674-11eb-80a9-1b3a5f2ebd9a.gif)

![GIF 3](https://user-images.githubusercontent.com/60740930/95021964-72631080-0674-11eb-8c35-90ccd46b42e6.gif)

### Test Coverage


- Server
    TextAssessmentResource - 93%
    AutomaticTextAssessmentConflictService - 91%
    TextAssessmentService - 90%

- Client Tests
<img width="1410" alt="Client coverage" src="https://user-images.githubusercontent.com/60740930/95795251-3110d780-0cea-11eb-8b02-e384771dedee.png">
